wallet/src/program_facades/amm.rs

```diff
@@ -0,0 +1,451 @@
+use common::{error::ExecutionFailureKind, rpc_primitives::requests::SendTxResponse};
+use nssa::{AccountId, program::Program};
+
+use crate::{
+    WalletCore,
+    cli::account::TokenHolding,
+    program_facades::{
+        OrphanHack49BytesInput, OrphanHack65BytesInput, compute_liquidity_token_pda,
+        compute_pool_pda, compute_vault_pda,
+    },
+};
+
+pub struct AMM<'w>(pub &'w WalletCore);
+
+impl AMM<'_> {
+    pub async fn send_new_amm_definition(
+        &self,
+        user_holding_a: AccountId,
+        user_holding_b: AccountId,
+        user_holding_lp: AccountId,
+        balance_a: u128,
+        balance_b: u128,
+    ) -> Result<SendTxResponse, ExecutionFailureKind> {
+        let (instruction, program) = amm_program_preparation_definition(balance_a, balance_b);
+
+        let amm_program_id = Program::amm().id();
+
+        let Ok(user_a_acc) = self.0.get_account_public(user_holding_a).await else {
+            return Err(ExecutionFailureKind::SequencerError);
+        };
+        let Ok(user_b_acc) = self.0.get_account_public(user_holding_b).await else {
+            return Err(ExecutionFailureKind::SequencerError);
+        };
+
+        let definition_token_a_id = TokenHolding::parse(&user_a_acc.data)
+            .ok_or(ExecutionFailureKind::AccountDataError(user_holding_a))?
+            .definition_id;
+        let definition_token_b_id = TokenHolding::parse(&user_b_acc.data)
+            .ok_or(ExecutionFailureKind::AccountDataError(user_holding_a))?
+            .definition_id;
+
+        let amm_pool =
+            compute_pool_pda(amm_program_id, definition_token_a_id, definition_token_b_id);
+        let vault_holding_a = compute_vault_pda(amm_program_id, amm_pool, definition_token_a_id);
+        let vault_holding_b = compute_vault_pda(amm_program_id, amm_pool, definition_token_b_id);
+        let pool_lp = compute_liquidity_token_pda(amm_program_id, amm_pool);
+
+        let account_ids = vec![
+            amm_pool,
+            vault_holding_a,
+            vault_holding_b,
+            pool_lp,
+            user_holding_a,
+            user_holding_b,
+            user_holding_lp,
+        ];
+
+        let Ok(nonces) = self
+            .0
+            .get_accounts_nonces(vec![user_holding_a, user_holding_b])
+            .await
+        else {
+            return Err(ExecutionFailureKind::SequencerError);
+        };
```

Comment:
```suggestion
        let nonces = self
            .0
            .get_accounts_nonces(vec![user_holding_a, user_holding_b])
            .await
            .map(|_| ExecutionFailureKind::SequencerError)?;
```

More rust way.

Some idea for other places